infra/blueprint-test/pkg/bq/bq.go

```diff
@@ -0,0 +1,165 @@
+	logger       *logger.Logger // custom logger
+}
+
+type cmdOption func(*CmdCfg)
+
+func WithBinary(bqBinary string) cmdOption {
+	return func(f *CmdCfg) {
+		f.bqBinary = bqBinary
+	}
+}
+
+func WithCommonArgs(commonArgs []string) cmdOption {
+	return func(f *CmdCfg) {
+		f.commonArgs = commonArgs
+	}
+}
+
+func WithLogger(logger *logger.Logger) cmdOption {
+	return func(f *CmdCfg) {
+		f.logger = logger
+	}
+}
+
+// newCmdConfig sets defaults and validates values for bq Options.
+func newCmdConfig(opts ...cmdOption) (*CmdCfg, error) {
+	gOpts := &CmdCfg{}
+	// apply options
+	for _, opt := range opts {
+		opt(gOpts)
+	}
+	if gOpts.bqBinary == "" {
+		err := utils.BinaryInPath("bq")
+		if err != nil {
+			return nil, err
+		}
+		gOpts.bqBinary = "bq"
+	}
+	if gOpts.commonArgs == nil {
+		gOpts.commonArgs = []string{"--format", "json"}
+	}
+	if gOpts.logger == nil {
+		gOpts.logger = utils.GetLoggerFromT()
+	}
+	return gOpts, nil
+}
+
+// RunCmd executes a bq command and fails test if there are any errors.
+func RunCmd(t testing.TB, cmd string, opts ...cmdOption) string {
+	op, err := RunCmdE(t, cmd, opts...)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return op
+}
+
+// RunCmdE executes a bq command and return output.
+func RunCmdE(t testing.TB, cmd string, opts ...cmdOption) (string, error) {
+	gOpts, err := newCmdConfig(opts...)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// split command into args
+	args := strings.Fields(cmd)
+	bqCmd := shell.Command{
+		Command: "bq",
+		Args:    append(args, gOpts.commonArgs...),
+		Logger:  gOpts.logger,
+	}
+	return shell.RunCommandAndGetStdOutE(t, bqCmd)
+}
+
+// Run executes a bq command and returns value as gjson.Result.
+// It fails the test if there are any errors executing the bq command or parsing the output value.
+func Run(t testing.TB, cmd string, opts ...cmdOption) gjson.Result {
+	op := RunCmd(t, cmd, opts...)
+	if !gjson.Valid(op) {
+		t.Fatalf("Error parsing output, invalid json: %s", op)
+	}
+	return gjson.Parse(op)
+}
+
+// RunWithCmdOptsf executes a bq command and returns value as gjson.Result.
+//
+// RunWithCmdOptsf(t, ops.., "ls --datasets --project_id=%s", "projectId")
+//
+// It fails the test if there are any errors executing the bq command or parsing the output value.
+func RunWithCmdOptsf(t testing.TB, opts []cmdOption, cmd string, args ...interface{}) gjson.Result {
+	return Run(t, stringFromTextAndArgs(append([]interface{}{cmd}, args...)...), opts...)
+}
+
+// Runf executes a bq command and returns value as gjson.Result.
+//
+// Runf(t, "ls --datasets --project_id=%s", "projectId")
+//
+// It fails the test if there are any errors executing the bq command or parsing the output value.
+func Runf(t testing.TB, cmd string, args ...interface{}) gjson.Result {
+	return Run(t, stringFromTextAndArgs(append([]interface{}{cmd}, args...)...))
+}
+
+// stringFromTextAndArgs convert msg and args to formatted text
+func stringFromTextAndArgs(msgAndArgs ...interface{}) string {
```

Comment:
Since we are starting to use this for both gcloud and bq, I think centralizing this in utils might be good.

Reply:
I can move this out of both files but that might result in a breaking change if the function in `gcloud.go` was referenced in any downstream projects - wdyt?

Reply:
Since it was unexported from gcloud package it is okay to move

Reply:
Works for me.

Reply:
Done!